### PR TITLE
REFACTOR: chat-channel-title

### PR DESCRIPTION
**Pull request by jjaffeux**
- tagLess
- fully tested

### CODE REVIEW

**Comment by nattsw** [comment]
i feel slightly torn about fabricators. on one hand it feels nice to just be able to create an object like that, but the alternative is that i have to look here and there to see what the defaults are, and sometimes it gets hard to modify one when it grows too big. also, if it’s some kind of unit / component test, i’d rather create the object in the ”unit” itself. if we’d have to reuse an object a lot in many places, it feels like a smell that maybe the component doesn’t exactly just do one thing and one thing well

File: assets/javascripts/discourse/models/chat-channel.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 export const CHATABLE_TYPES = {
   directMessageChannel: "DirectMessageChannel",
   topicChannel: "Topic",
+  tagChannel: "Tag",
```
Comment (nattsw) [inline assets/javascripts/discourse/models/chat-channel.js:4]: I recall @markvanlan saying that tagChannels will be gone soon.

**Comment by jjaffeux** [comment]
> alternative is that i have to look here and there to see what the defaults are

No silver bullet on this, but if you don't do this, you would also have to look what are the expected properties

> and sometimes it gets hard to modify one when it grows too big

yes maybe? that's speculation, and I would rather have one slightly complicated object to change, than having to change dozens of custom objects everywhere in my tests when I change one thing in my model

> if it’s some kind of unit / component test, i’d rather create the object in the ”unit” itself.  if we’d have to reuse an object a lot in many places, it feels like a smell that maybe the component doesn’t exactly just do one thing and one thing well

The alternative to this, is passing only exactly what you need, I feel like it often ends up in gigantic list of properties, is hard to maintain, generates lot of code and make changes in nested components complicated as you have a lot of things to change in the hierarchy, passing the channel property to a component showing the title of a channel doesn't feel like a smell to me

File: assets/javascripts/discourse/models/chat-channel.js (hunk 1 above)
Comment (jjaffeux) [reply assets/javascripts/discourse/models/chat-channel.js:4]: Yes I think you are rigjt. I guess we can remove this when they are actually gone.

**Comment by nattsw** [comment]
> that’s speculation 
I don’t think it’s speculation. Take our chat-fixtures file for example. It has grown too big and clunky that when I try to modify one thing, something else breaks.

> gigantic list of properties

If a component is well segregated (aka SRP), we should not have to pass in that many attributes. As an example, when I looked at the chat channel title component, I expect it to be invoked like this `{{chat-channel-title title=channel.title}}` because why should the title component take in an entire channel be aware of it?

But I understand, the refactor is okay. What we’re talking about is something bigger than this PR. Though I would still prefer not to have the fabricator.

**Review by nattsw — APPROVED** [review]
I think overall it's okay. I don't want differing *opinions* to get in the way.

**Comment by jjaffeux** [comment]
> If a component is well segregated (aka SRP), we should not have to pass in that many attributes. As an example, when I looked at the chat channel title component, I expect it to be invoked like this `{{chat-channel-title title=channel.title}}` because why should the title component take in an entire channel be aware of it?
> 
> But I understand, the refactor is okay. What we’re talking about is something bigger than this PR. Though I would still prefer not to have the fabricator.

I really don't agree on this, it's not because it's called `{{chat-channel-title}}` that it's only property should be a title... if you go this way you will need A LOT of components. While I don't think this is currently perfect I think this is totally ok, a simple proof of this is that it's very easy to test.

Having only to create one single chat-channel object to test a component is rather lightweight IMO.

**Comment by jjaffeux** [comment]
> I don’t think it’s speculation. Take our chat-fixtures file for example. It has grown too big and clunky that when I try to modify one thing, something else breaks. There’s also evidence of this in core.

I'm sorry... but that doesn’t prove anything. First it's fixtures not fabricators, if you want evidence of fabricators in core, we have theme everywhere and they work quite well. And then I stand by my point that's it's speculation because, I refused to keep creating small channel objects with always the same copy-pasted params and created this small helper.

Wether it will become some gigantic mess is not related to fabricators, but to future developments of this. My point being it has more chance to become a gigantic mess if we keep creating lots of small objects everywhere than trying to DRY it now.

**Pull request merged**